dask/dataframe/core.py

```diff
@@ -309,6 +310,261 @@ def _scalar_binary(op, self, other, inv=False):
+        # include: "ascending", "descending" and "hash".
+        #   e.g. {<column-name-tuple>: <how-str>}
+        # This is distinct from `divisions`, which requires
+        # sorted ordering to be useful.
+        # NOTE: "__index__" corresponds to an unnamed index.
+        if isinstance(partitioning, dict):
+            self._partitioning = partitioning
+        else:
+            self._partitioning = {("__index__",): "order"} if divisions else {}
+
+        # Optional partition-wise statistics:
+        #   - "__num_rows__": Series of partition row counts
+        #   - <column-name>: DataFrame of partition stats for the column
+        #     - Required DataFrame columns: "min" and "max"
+        if isinstance(statistics, PartitionStatistics):
+            self._statistics = statistics
+        else:
+            self._statistics = PartitionStatistics(statistics)
+
+    def copy(self, **kwargs):
+        """Return copy of this PartitionMetadata object"""
+        new_kwargs = dict(
+            meta=None if self.meta is None else self.meta.copy(),
+            npartitions=self.npartitions,
+            divisions=self.divisions,
+            partitioning=self.partitioning.copy(),
+            statistics=self._statistics.copy(),
+        )
+        new_kwargs.update(**kwargs)
+        return PartitionMetadata(**new_kwargs)
+
+    def __getstate__(self):
+        return dict(
+            _meta=None if self.meta is None else self.meta.copy(),
+            _npartitions=self.npartitions,
+            _divisions=self.divisions,
+            _partitioning=self.partitioning.copy(),
+            _statistics=self._statistics.copy(),
+        )
+
+    @property
+    def meta(self) -> Any:
+        """Return global DataFrame schema"""
+        return self._meta
+
+    def set_meta(self, value) -> None:
+        """Set global DataFrame schema"""
+        self._meta = value
+
+    @property
+    def npartitions(self) -> int:
+        """Total partition count"""
+        return self._npartitions
+
+    @lru_cache(maxsize=1)
+    def _unknown_divisions(self, npartitions) -> tuple:
+        """Return conventional 'unknown' divisions"""
+        return (None,) * (npartitions + 1)
+
+    @property
+    def divisions(self) -> tuple:
+        """Get Min/max Index statistics, if known"""
+        if not self.known_divisions:
+            return self._unknown_divisions(self.npartitions)
+        if isinstance(self._divisions, (list, tuple)):
+            return tuple(self._divisions)
+        else:
+            raise ValueError
+
+    @property
+    def known_divisions(self):
+        """Return if divisions are known"""
+        return self._divisions is not None
+
+    def set_divisions(self, value: tuple):
+        """Set Min/max Index statistics"""
+        if isinstance(value, (list, tuple)) and all(map(lambda x: x is None, value)):
+            self._divisions = None
+        else:
+            self._divisions = value
```

Comment:
What would you think about removing `set_divisions`, `set_meta`, and `set_partitioning` all-together, and making everything going through `copy` (or similar)? As with some of our other discussions, I'd love to see the representation of dataframe metadata move in a direction towards immutability.